source/module_hamilt_lcao/module_gint/gint_vl_gpu.cu

```diff
@@ -36,311 +36,208 @@ namespace GintKernel
 
-                if (lo1 <= lo2)
+            if (lo1 <= lo2)
+            {
+                const hamilt::AtomPair<double>* tmp_ap
+                    = hRGint->find_pair(iat1, iat2);
+                if (tmp_ap == nullptr)
                 {
-                    hamilt::AtomPair<double>* tmp_ap
-                        = hRGint->find_pair(iat1, iat2);
-                    if (tmp_ap == nullptr)
-                    {
-                        continue;
-                    }
-                    int atom_pair_nw
-                        = ucell.atoms[it1].nw * ucell.atoms[it2].nw;
-                    if (gridt.grid_vlocal_g[iat1 * ucell.nat + iat2] == nullptr)
-                    {
-                        checkCuda(cudaMallocAsync(
-                            (void**)&gridt
-                                .grid_vlocal_g[iat1 * ucell.nat + iat2],
-                            atom_pair_nw * sizeof(double),
-                            gridt.streams[stream_num]));
-                    }
-                    checkCuda(cudaMemsetAsync(
-                        gridt.grid_vlocal_g[iat1 * ucell.nat + iat2],
-                        0,
-                        atom_pair_nw * sizeof(double),
-                        gridt.streams[stream_num]));
-                    iter_num++;
+                    continue;
                 }
+                const int atom_pair_nw
+                    = ucell.atoms[it1].nw * ucell.atoms[it2].nw;
+                grid_vlocal_g[iat1 * ucell.nat + iat2] = 
+                    Cuda_Mem_Wrapper<double>(atom_pair_nw, 1, false);
+                grid_vlocal_g[iat1 * ucell.nat + iat2].memset_device_sync();
             }
         }
     }
-    for (int i = 0; i < gridt.nstreams; i++)
-    {
-        checkCuda(cudaStreamSynchronize(gridt.streams[i]));
-    }
 
-#pragma omp parallel for num_threads(gridt.nstreams) collapse(2)
+    Cuda_Mem_Wrapper<double> psi_input_double(5 * max_atom_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> psi_input_int(2 * max_atom_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> atom_num_per_bcell(nbzp, num_streams, true);
+
+    Cuda_Mem_Wrapper<double> psi(max_phi_per_z, num_streams, false);
+    Cuda_Mem_Wrapper<double> psi_vldr3(max_phi_per_z, num_streams, false);
+
+    Cuda_Mem_Wrapper<int> gemm_m(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> gemm_n(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> gemm_k(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> gemm_lda(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> gemm_ldb(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<int> gemm_ldc(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<double*> gemm_A(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<double*> gemm_B(max_atompair_per_z, num_streams, true);
+    Cuda_Mem_Wrapper<double*> gemm_C(max_atompair_per_z, num_streams, true);
```

Comment:
gemm_{m,n,k,lda,ldb,ldc,A,B,C},psi_input_{double,int} can use a structure to store it, making it convenient for the three calculations.

Reply:
OK, maybe we can fix it next time